### PR TITLE
Add proxy-bypass option to prevent using proxy depending on rule

### PR DESCRIPTION
**Pull request by AlexKamaev**

### CODE REVIEW

**Comment by AlexKamaev** [comment]
see API in tests

File: test/server/match-url-test.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
+var expect   = require('chai').expect;
+var matchUrl = require('../../lib/utils/check-url');
+
+it('Should fallback to the default reporter if reporter was not set', function () {
```
Comment (helen-dikareva) [inline test/server/match-url-test.js]: Strange name for the test

**Comment by AlexKamaev** [comment]
@testcafe-build-bot \retest

File: src/cli/argument-parser.js [hunk 2]
```diff
@@ -107,6 +107,7 @@ export default class CLIArgumentParser {
             .option('--ports <port1,port2>', 'specify custom port numbers')
             .option('--hostname <name>', 'specify the hostname')
             .option('--proxy <host>', 'specify the host of the proxy server')
+            .option('--proxy-bypass <url1,url2>', 'specify the urls to bypass the proxy server')
```
Comment (helen-dikareva) [inline src/cli/argument-parser.js]: We need to check type of url template value and create error for a case when we cannot parse it

File: test/server/match-url-test.js [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    expect(matchUrl('http://localhost', rule)).to.be.true;
+    expect(matchUrl('my-localhost', rule)).to.be.false;
+    expect(matchUrl('localhost-my', rule)).to.be.false;
+
+    rule = ['127.0.0.1'];
+
+    expect(matchUrl('127.0.0.1', rule)).to.be.true;
+    expect(matchUrl('http://127.0.0.1', rule)).to.be.true;
+    expect(matchUrl('https://127.0.0.1', rule)).to.be.true;
+    expect(matchUrl('https://127.0.0.1', rule)).to.be.true;
```
Comment (AndreyBelym) [inline test/server/match-url-test.js]: Duplicate?

Comment (AlexKamaev) [reply test/server/match-url-test.js]: yes. mistake

File: test/server/match-url-test.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
+var expect   = require('chai').expect;
+var matchUrl = require('../../lib/utils/check-url');
+
+it('Should check does url match rule', function () {
+    var rule = ['google.com'];
```
[the remark this answers is not in the document]
Comment (AlexKamaev) [reply test/server/match-url-test.js]: yes we can

File: src/cli/argument-parser.js (hunk 2 above)
Comment (AlexKamaev) [reply src/cli/argument-parser.js]: I thought it is not required to create error. If rule is incorrent test will run via proxy, and that's all. Rule just will be ignored. Don't you agree?

Comment (VasilyStrelyaev) [inline src/cli/argument-parser.js]:
```
            .option('--proxy-bypass <rules>', 'specify a comma-separated list of rules that define URLs accessed bypassing the proxy server')
```

Comment (VasilyStrelyaev) [reply src/cli/argument-parser.js]: Regarding https://github.com/DevExpress/testcafe/pull/2187#discussion_r172121922 we can just add `Rules consist of URLs with wildcards enabled`.
Is that simpler. @AndreyBelym ?

**Comment by AlexKamaev** [comment]
@testcafe-build-bot \retest

**Comment by AlexKamaev** [comment]
FPR

**Comment by n0v1** [comment]
Great to see this feature beeing added! Thank you! 

Is this expected to work for ressources like CSS stylesheets or JavaScript files that are linked from the page under test?

I'm sitting behind a corporate proxy and want to test a page that is served by a web server on localhost. The page loads a JavaScript library from a remote host. So the request to load the page should bypass the proxy, the request to get the library should go through the proxy.

I call TestCafé CLI with `--proxy 172.30.109.30:8080 --proxy-bypass 127.0.0.1:8000`. The page and all local ressources load correctly but the remote script does not. It keeps loading forever until a timeout is hit eventually.

To test in isolation I created this small repo: [n0v1/testcafe-proxy-bypass-test](/n0v1/testcafe-proxy-bypass-test)

**Comment by AlexKamaev** [comment]
Hi
Now this feature is implemented only to bypass proxy for specific tests according to the test page url. But your case seems to be important, so I think we'll support it as well.

**Comment by amirse80** [comment]
Leave a comment

**Comment by AlexKamaev** [comment]
Hello @amirse80 
TestCafe CLI supports two options for using proxy. Please refer to the following articles:
[--proxy](http://devexpress.github.io/testcafe/documentation/using-testcafe/command-line-interface.html#--proxy-host) and [--proxy-bypass](http://devexpress.github.io/testcafe/documentation/using-testcafe/command-line-interface.html#--proxy-bypass-rules).
According to our documentation, the `proxy`option specifies the proxy server used in your local network to access the Internet. If you do not have access to the second URL, I suppose that you need to use the `proxy` option.
The `proxy-bypass` option can be used together with the `proxy` option. It's required when you use proxy, but don't need to use proxy only for some specific resources.

**Comment by amirse80** [comment]
Hello,
Thank you for the input. Unfortunatelly it didn´t solve my problem.
The issue is like this.
1. I navigate to an company internal page.    (.page `http://dap-internal.company.page´)
2. The chrome is started and initially navigates to the correct server.
3. During the page opening the URL changes, redirects, to **http://pen-internal.company.page**. Now this new URL works fine in the same browser that testcafe has started IF i copy the real URL manually into the new tab BUT testcafe ADDS **http://172.30.48.47:61019/DSc6im7OL/** to the URL which makes it not working. So instead of the real redirected URL **http://pen-internal.company.page** the redirected URL is **http://172.30.48.47:61019/DSc6im7OL/http://pen.internal.company.page**!

**Pull request merged**